File: android/domain/domain-auth/src/main/kotlin/app/kehdo/domain/auth/SignInUseCase.kt
package app.kehdo.domain.auth

import app.kehdo.core.common.Outcome
import javax.inject.Inject

/**
 * Sign in with email + password.
 *
 * Use-cases are single-purpose, suspend, return Outcome<T>.
 * They contain orchestration logic only — no HTTP, no DB.
 */
class SignInUseCase @Inject constructor(
    private val authRepository: AuthRepository
) {
    suspend operator fun invoke(email: String, password: String): Outcome<User> {
        require(email.isNotBlank()) { "Email cannot be blank" }
        require(password.length >= 8) { "Password must be at least 8 characters" }
        return authRepository.signIn(email.trim().lowercase(), password)
    }
}

File: ios/Packages/KHDomain/Sources/KHDomain/SignInUseCase.swift
import Foundation
import KHCore

/// Sign in with email + password.
///
/// Use-cases are single-purpose, async, return `Outcome<T>`.
/// They contain orchestration logic only — no networking or persistence.
public struct SignInUseCase: Sendable {
    private let authRepository: AuthRepository

    public init(authRepository: AuthRepository) {
        self.authRepository = authRepository
    }

    public func callAsFunction(email: String, password: String) async -> Outcome<User> {
        let trimmed = email.trimmingCharacters(in: .whitespaces).lowercased()
        guard !trimmed.isEmpty else {
            return .failure(.server(code: "INVALID_EMAIL", message: "Email cannot be blank"))
        }
        guard password.count >= 8 else {
            return .failure(.server(code: "INVALID_PASSWORD", message: "Password must be at least 8 characters"))
        }
        return await authRepository.signIn(email: trimmed, password: password)
    }
}
